autoprecompiles/src/low_degree_bus_interaction_optimizer.rs

```diff
@@ -0,0 +1,554 @@
+                    symbolic_function.output.expression.clone() - low_degree_function;
+
+                // Check degree
+                let within_degree_bound =
+                    polynomial_constraint.degree() <= self.degree_bound.identities;
+                within_degree_bound.then_some(polynomial_constraint)
+            })
+    }
+
+    /// Given a bus interaction of 2 or 3 unknown fields, finds all combinations of (symbolic)
+    /// inputs and outputs where the input space is small enough.
+    fn symbolic_functions_with_small_domain(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Vec<SymbolicFunction<T, V>> {
+        let expressions = bus_interaction
+            .payload
+            .iter()
+            .enumerate()
+            .filter(|(_i, expr)| expr.try_to_number().is_none())
+            .map(|(i, expr)| (i, expr.clone()))
+            .collect::<BTreeMap<_, _>>();
+        let range_constraints = expressions
+            .iter()
+            .map(|(i, expr)| (*i, self.solver.range_constraint_for_expression(expr)))
+            .collect::<BTreeMap<_, _>>();
+        let unknown_field_indices = expressions.keys().copied().collect::<Vec<_>>();
+
+        // Currently, we only have hypotheses for:
+        // - 2 unknown fields (1 input, 1 output)
+        // - 3 unknown fields (2 inputs, 1 output)
+        if !(unknown_field_indices.len() == 2 || unknown_field_indices.len() == 3) {
+            return Vec::new();
+        }
+
+        unknown_field_indices
+            .iter()
+            .copied()
+            .combinations(unknown_field_indices.len() - 1)
+            .filter(|input_indices| {
+                self.has_small_domain(input_indices.iter().map(|i| range_constraints[i].clone()))
+            })
+            // Build the input-output pairs
+            .map(|input_indices| {
+                let output_index = unknown_field_indices
+                    .iter()
+                    .cloned()
+                    .filter(|i| !input_indices.contains(i))
+                    .exactly_one()
+                    .unwrap();
+                SymbolicFunction {
+                    inputs: input_indices
+                        .iter()
+                        .map(|i| SymbolicInput {
+                            index: *i,
+                            expression: expressions[i].clone(),
+                            range_constraint: range_constraints[i].clone(),
+                        })
+                        .collect(),
+                    output: SymbolicOutput {
+                        index: output_index,
+                        expression: expressions[&output_index].clone(),
+                    },
+                }
+            })
+            .collect()
+    }
+
+    /// Given a list of input range constraints, computes whether the total input space (i.e.,
+    /// the cross product of all possible input values) is small enough.
+    fn has_small_domain(
```

Comment:
Maybe just call it `has_few_possible_values`? The parameter is not a function and the name of the parameter is also not obvious at the call site.